### PR TITLE
Replace PhantomJS with ChromeHeadless for unit test support

### LINKED ISSUE #9276: Move from PhantomJS to Chrome Headless for client tests
PhantomJS is giving us a lot of problem due to the fact that it's using an outdated JS engine.

We should move to Chrome Headless for this.

In a PR that's currently in progress but that's going to be merged soon we had to disable unit tests because of this, so when this issue is fixed client tests should be enabled again by changing this line in package.json (under "scripts")

```
"test": "npm run lint && gulp test && gulp apidoc",
```

to 

```
"test": "npm run lint && gulp test && npm run client:unit && gulp apidoc",
```

Important links:
- https://github.com/HabitRPG/habitica/blob/develop/test/client/unit/karma.conf.js#L19
- https://github.com/karma-runner/karma-chrome-launcher
- https://github.com/HabitRPG/habitica/pull/9236/commits/50c156d9d047198a7fb9bffc90c6069bfb456f12

### PR DESCRIPTION
**Pull request by kartikadur**
[//]: # (Note: See http://habitica.wikia.com/wiki/Using_Your_Local_Install_to_Modify_Habitica%27s_Website_and_API for more info)

[//]: # (Put Issue # or URL here, if applicable. This will automatically close the issue if your PR is merged in)
Fixes #9276 

### Changes
[//]: # (Describe the changes that were made in detail here. Include pictures if necessary)
#### karma.conf.js
+ browser from PhantomJS to ChromeHeadless
#### package.json (package.lock.json by extension)
+ remove phantomjs-prebuilt
+ remove karma-phantomjs-launcher
+ add karma-chrome-launcher
#### .travis.yml
+ chrome addon support
+ unit test support

_todo: add docker support for chrome browser_

[//]: # (Put User ID in here - found on the Habitica website at User Icon > Settings > API)

----
UUID:

### CODE REVIEW

**Comment by kartikadur** [comment]
chrome-headless integration is complete and works. 

unit tests run successfully. the two errors are due to some problem in the test itself and may need to be modified.

Not sure why the API_V3 tests failed.

File: .travis.yml [hunk 1]
```diff
@@ -1,3 +1,4 @@
+dist: trusty
```
Comment (paglias) [inline .travis.yml]: can we avoid specifying the dist?

**Review by paglias — COMMENTED** [review]
Thanks a lot for this! We'll have to get the tests passing before we can merge this, though. Don't worry about the random failure in API_V3.

Reading https://github.com/karma-runner/karma-chrome-launcher it seems that this requires chrome to be preinstalled on the platform where you're running tests so these won't work with Docker. Can we add puppeteer as explained here https://github.com/karma-runner/karma-chrome-launcher#headless-chromium-with-puppeteer?

File: .travis.yml (hunk 1 above)
Comment (kartikadur) [reply .travis.yml]: I added the dist command based on documentation by travis CI. Its seems that linux 12 has problems with chrome (headless) and this command ensure that linux 14 is used. If linux 14 is used by default I can remove it.

**Comment by kartikadur** [comment]
If puppeteer works then I could even get rid of the dist command in travis.yml. I'll look into it.

File: .travis.yml (hunk 1 above)
Comment (paglias) [reply .travis.yml]: pretty sure that trusty is the default now (we upgraded a few weeks ago)

Comment (kartikadur) [reply .travis.yml]: ok good to know, will remove it.

**Comment by kartikadur** [comment]
For the moment I have skipped the two failing tests so the chrome-headless update at the very least passes. I'll add the two failing tests in another PR, once I figure out why they are failing.

**Comment by paglias** [comment]
Hmm not sure why but the integration tests seems to be broken in this PR, I restarted the build 3 times but still failing

**Comment by paglias** [comment]
Hmm not sure why but the integration tests seems to be broken in this PR, I restarted the build 3 times but still failing

**Comment by paglias** [comment]
According to this mongoose issue https://github.com/Automattic/mongoose/issues/4490

try this 

```
--- a/test/helpers/api-unit.helper.js
+++ b/test/helpers/api-unit.helper.js
@@ -11,7 +11,7 @@ import * as Tasks from '../../website/server/models/task';

 afterEach((done) => {
   sandbox.restore();
-  mongoose.connection.db.dropDatabase(done);
+  mongoose.connection.dropDatabase(done);
 });

 export { sleep } from './sleep';
diff --git a/test/helpers/mongo.js b/test/helpers/mongo.js
index 94ce48adc..fb16e3a24 100644
--- a/test/helpers/mongo.js
+++ b/test/helpers/mongo.js
@@ -37,7 +37,7 @@ export async function getProperty (collectionName, id, path) {
 // resets the db to an empty state and creates a tavern document
 export async function resetHabiticaDB () {
   return new Promise((resolve, reject) => {
-    mongoose.connection.db.dropDatabase((dbErr) => {
+    mongoose.connection.dropDatabase((dbErr) => {
       if (dbErr) return reject(dbErr);
       let groups = mongoose.connection.db.collection('groups');
       let users = mongoose.connection.db.collection('users');
@@ -119,7 +119,7 @@ before((done) => {
 });

 after((done) => {
-  mongoose.connection.db.dropDatabase((err) => {
+  mongoose.connection.dropDatabase((err) => {
:
```

**Comment by paglias** [comment]
Upgraded some deps in the `develop` branch

**Comment by paglias** [comment]
Upgraded some deps in `develop`, you might want to merge into your branch

**Comment by kartikadur** [comment]
:tada:

**Comment by paglias** [comment]
I'm a bit busy today, but tomorrow I'll test it in docker and if it works we'll merge it

**Comment by kartikadur** [comment]
sorry, the :tada: was cause it passed without issues. thanks.





*Kartik K AdurWeb/UI Developer, Information Architect, Interaction
DesignerLinkedIn: Kartik Adur Public Profile
<https://www.linkedin.com/pub/kartik-adur/8/b5/894>Personal Website:
kartikadur.wordpress.com <http://kartikadur.wordpress.com>Email:
kartikadur@gmail.com <kartikadur@gmail.com>*

**Comment by paglias** [comment]
I'm having some issues setting up Docker and testing this, can you try @SabreCat @TheHollidayInn ? Not a problem with the PR but I can't find out how to run tests on docker

**Comment by TheHollidayInn** [comment]
I don't see anything in the PR that is docker dependent. What needs to be tested with Docker?

**Comment by paglias** [comment]
I was trying to see it running client tests with docker worked. But npm install worked so we should be fine since that also installs chrome

**Comment by paglias** [comment]
@kartikadur  thanks for this! I'm merging the PR if you post your UUID here I'll award your contributor tier on Habitica.

**Pull request merged**